nomad/task_group_host_volume_claim_endpoint.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package nomad
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-memdb"
+	metrics "github.com/hashicorp/go-metrics/compat"
+	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/nomad/state"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// TaskGroupHostVolumeClaim is the server RPC endpoint for task group volume claims
+type TaskGroupHostVolumeClaim struct {
+	srv    *Server
+	ctx    *RPCContext
+	logger hclog.Logger
+}
+
+func NewTaskGroupVolumeClaimEndpoint(srv *Server, ctx *RPCContext) *TaskGroupHostVolumeClaim {
+	return &TaskGroupHostVolumeClaim{srv: srv, ctx: ctx, logger: srv.logger.Named("task_group_host_volume_claim")}
+}
+
+func (tgvc *TaskGroupHostVolumeClaim) List(args *structs.TaskGroupVolumeClaimListRequest, reply *structs.TaskGroupVolumeClaimListResponse) error {
+	authErr := tgvc.srv.Authenticate(tgvc.ctx, args)
+	if done, err := tgvc.srv.forward(structs.TaskGroupHostVolumeClaimListRPCMethod, args, args, reply); done {
+		return err
+	}
+	tgvc.srv.MeasureRPCRate("task_group_volume_claim", structs.RateMetricList, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "task_group_volume_claim", "list"}, time.Now())
+
+	// TODO: should this be a separate ACL capability? (nooo?)
+	allowClaim := acl.NamespaceValidator(acl.NamespaceCapabilityHostVolumeRead)
+	aclObj, err := tgvc.srv.ResolveACL(args)
+	if err != nil {
+		return err
+	}
+	if !allowClaim(aclObj, args.RequestNamespace()) {
+		return structs.ErrPermissionDenied
+	}
+
+	return tgvc.srv.blockingRPC(&blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, stateStore *state.StateStore) error {
+
+			// The iteration below appends directly to the reply object, so in
+			// order for blocking queries to work properly we must ensure the
+			// Claims are reset. This allows the blocking query run function to
+			// work as expected.
+			reply.Claims = nil
+
+			iter, err := stateStore.GetTaskGroupHostVolumeClaims(ws)
+			if err != nil {
+				return err
+			}
+
+			// Iterate all the results and add these to our reply object.
+			for raw := iter.Next(); raw != nil; raw = iter.Next() {
+				reply.Claims = append(reply.Claims, raw.(*structs.TaskGroupHostVolumeClaim))
+			}
```

Comment:
We're missing pagination and bexpr filtering, as well as ACL filtering for the case where the user has provided a wildcard namespace.

Reply:
[21c8351](https://github.com/hashicorp/nomad/pull/25114/commits/21c8351c169557ffc1b8440b038c743e37098987)